### PR TITLE
loader: stop retry execute sqls when closed 

### PR DESCRIPTION
**Pull request by WangXiangUSTC**
### What problem does this PR solve? 
fix issue: https://internal.pingcap.net/jira/browse/TOOL-1018
when execute sqls meet error and user kill drainer, drainer will not exit immediately and will still retry util reach the max retry time.

### What is changed and how it works?
1. add stopRetry function as argument in RetryOnError function
2. when loader is closed, stop retry

### Check List 

Tests 

 - Unit test
 - Manual test (when drainer execute sql and meet error, kill drainer, check drainer is exit or not)

Related changes

 - Need to cherry-pick to the release branch

### CODE REVIEW

**Comment by kennytm** [comment]
Why not use `RetryContext` (and cancel a Context on exit)?

**Comment by WangXiangUSTC** [comment]
do you mean add a context in `loader`, and judge ctx.Done() when retry? But it is not suggest to store contexts inside a struct.  @kennytm

**Comment by kennytm** [comment]
Besides storing the context into the struct, execDDL could accept a context as argument.

**Comment by WangXiangUSTC** [comment]
it is difficult to use context as argument in this situation now @kennytm

**Comment by WangXiangUSTC** [comment]
if so we need add a context in loaderImpl, do you think it's ok? @suzaku

**Comment by suzaku** [comment]
> if so we need add a context in loaderImpl, do you think it's ok? @suzaku

I think it's OK in this case because it could enable the reuse of `RetryContext` so that we don't need to change `RetryOnError`.

**Comment by WangXiangUSTC** [comment]
code updated, but I think we may need refine it someday, it is not suggest to store context in struct. @kennytm @suzaku PTAL again

**Comment by kennytm** [comment]
Hmm I don't know, what I had in mind is make it `execDDL(Context, *DDL)` up to `(*loaderImpl) Run(Context)`, and when the program stops the Context is canceled externally.

**Comment by WangXiangUSTC** [comment]
drainer/server -> drainer/syncer -> drainer/sync/MysqlSyncer -> pkg/Loader, so we need add context in MysqSyncer? or drainer/server?  and loader already provide a Close() function, why need do cancel context again? @kennytm

**Comment by kennytm** [comment]
@WangXiangUSTC basically add Context to every function 🤪. So you don't need to store Context in the struct. Implementation-wise LGTM but I'm not sure about the design. WDYT @suzaku

**Comment by suzaku** [comment]
@kennytm I think these models as a whole were not designed to work well with context, maybe we can improve this with another PR later. For now, allowing `loaderImpl` to manage its own lifetime `context` should be the solution that requires minimal change.

**Comment by suzaku** [comment]
I just tried to refactor these functions to make them `context`-aware, but then I noticed that I also had to refactor the caller of `Loader`, it would be a lot of work.
A bigger problem is that we didn't design the `Close` methods very well in the first place, so in many cases, many cooperation need to happen for both the caller and the callee to quit correctly.

BTW, @WangXiangUSTC , there are many `Exec` calls under the hood, shall we also replace them with `ExecContext`?

**Comment by WangXiangUSTC** [comment]
@suzaku updated in https://github.com/pingcap/tidb-binlog/pull/691/commits/694765280940c2825c0043c2f31071e5d05a8f64 and https://github.com/pingcap/tidb-binlog/pull/691/commits/076a4558c46cd1c56b32be3867d6b970e83b859d

File: pkg/loader/load.go [hunk 1]
```diff
@@ -86,6 +86,9 @@ type loaderImpl struct {
 	// value can be tidb or mysql
 	saveAppliedTS           bool
 	lastUpdateAppliedTSTime time.Time
+
+	ctx    context.Context
```
Comment (kennytm) [inline pkg/loader/load.go]: Add a TODO that we shouldn't store the `ctx` in the struct and should be refactored in the future etc.

File: pkg/loader/load.go [hunk 2]
```diff
@@ -222,6 +232,7 @@ func (s *loaderImpl) Successes() <-chan *Txn {
 // Run will quit when all data is drained
 func (s *loaderImpl) Close() {
```
Comment (july2993) [inline pkg/loader/load.go]:
```
221 // Close close the Loader, no more Txn can be push into Input()
222 // Run will quit when all data is drained
223 func (s *loaderImpl) Close() {
224     close(s.input)
225 }
```
what's the behavior now?

File: pkg/loader/load.go [hunk 3]
```diff
@@ -222,6 +232,7 @@ func (s *loaderImpl) Successes() <-chan *Txn {
...
 	close(s.input)
+	s.cancel()
```
Comment (july2993) [inline pkg/loader/load.go]: in `Run()` it will still execute all the accumulated dmls (but with the context that has been canceled?)

Comment (WangXiangUSTC) [reply pkg/loader/load.go]: already remove ctx when execute sqls

**Comment by WangXiangUSTC** [comment]
@july2993 @kennytm PTAL again

File: pkg/loader/load_test.go [hunk 4]
```diff
@@ -58,6 +59,7 @@ func (cs *LoadSuite) TestGetExecutor(c *check.C) {
 		metrics: &MetricsGroup{
 			QueryHistogramVec: &prometheus.HistogramVec{},
 		},
+		ctx: context.Background(),
```
Comment (kennytm) [inline pkg/loader/load_test.go]: The `cancel` field needs to be non-null to avoid `s.cancel()` crashing, right?

Comment (WangXiangUSTC) [reply pkg/loader/load_test.go]: yes

Comment (WangXiangUSTC) [reply pkg/loader/load_test.go]: but the unit test dosen't call loader.Close(), no crashing now

**Comment by july2993** [comment]
@suzaku PTAL and approve it if still LGTM for you

**Pull request merged**